narwhals/_arrow/dataframe.py

```diff
@@ -615,3 +615,45 @@ def sample(
         mask = rng.choice(idx, size=n, replace=with_replacement)
 
         return self._from_native_frame(pc.take(frame, mask))
+
+    def unpivot(
+        self: Self,
+        on: str | list[str] | None,
+        index: str | list[str] | None,
+        variable_name: str | None,
+        value_name: str | None,
+    ) -> Self:
+        import pyarrow as pa  # ignore-banned-import
+
+        native_frame = self._native_frame
+        variable_name = variable_name if variable_name is not None else "variable"
+        value_name = value_name if value_name is not None else "value"
+
+        index_: list[str] = (
+            [] if index is None else [index] if isinstance(index, str) else index
+        )
+        on_: list[str] = (
+            [c for c in self.columns if c not in index_]
+            if on is None
+            else [on]
+            if isinstance(on, str)
+            else on
+        )
+
+        n_rows = len(self)
+
+        return self._from_native_frame(
+            pa.concat_tables(
+                [
+                    pa.Table.from_arrays(
+                        [
+                            *[native_frame.column(idx_col) for idx_col in index_],
+                            pa.array([on_col] * n_rows, pa.string()),
+                            native_frame.column(on_col),
+                        ],
+                        names=[*index_, variable_name, value_name],
+                    )
+                    for on_col in on_
+                ]
+            )
+        )
```

Comment:
Adapted from the following [StackOverflow Answer](https://stackoverflow.com/a/71660765/12411536)